# **Messi outlier**
## DATASET 2
### "2.2M+ Records -Most Comprehensive Football"

## Exploratory analyses of `player_performances.csv`

`player_performances.csv` is part of a collection called "2.2M+ Records -Most Comprehensive Football" (salimt). The collection is available in [Kaggle](https://www.kaggle.com/datasets/xfkzujqjvx97n/football-datasets).

`player_performances.csv` is a vast list of football players performances by season.

The aim of this notebook is to examine if `player_performances.csv` is suitable to compare Messi performance with other players.

From now on, we use `.parquet` files. `player_performances.csv` was previously transformed to `dataset2_player_performances.parquet`.

#### Route to `player_performances.parquet`

In [5]:
# This chunk is to set the path from the local folder to the player_performances.parquet in the data folder
from pathlib import Path

# project root is one level up from notebooks/
ROOT = Path.cwd().parent
DATA = ROOT / "data"

file_path = DATA / "dataset2_player_performances.parquet"

#### Load `appearances_FULL.parquet`, hereafter `df`.

In [6]:
# Load the file
import pandas as pd

# Load the parquet file
df = pd.read_parquet(file_path)

#### Exploratory analyses

In [11]:
# Basic info 
print("Shape:", df.shape)         # rows, columns
print(df.dtypes)                  # types of each column

# Since I want to know time span of the `df` I summarize `season_name` variable

if "season_name" not in df.columns:
    raise KeyError("season_name column not found in dataset")

season_min = df["season_name"].min()
season_max = df["season_name"].max()
unique_seasons = df["season_name"].nunique()
most_freq = df["season_name"].value_counts().idxmax()
most_freq_count = df["season_name"].value_counts().max()

print("📊 season_name summary")
print(f"• Season span: {season_min} → {season_max}")
print(f"• Number of unique seasons: {unique_seasons}")
print(f"• Most frequent season: {most_freq} ({most_freq_count:,} rows)")

# Optional: show counts per season
season_counts = df["season_name"].value_counts().sort_index()
print("\nCounts per season:")
display(season_counts)           

Shape: (760125, 20)
player_id                int64
season_name             object
competition_id          object
competition_name        object
team_id                  int64
team_name               object
nb_in_group              int64
nb_on_pitch              int64
goals                  float64
assists                  int64
own_goals                int64
subed_in                 int64
subed_out                int64
yellow_cards             int64
second_yellow_cards      int64
direct_red_cards         int64
penalty_goals            int64
minutes_played         float64
goals_conceded           int64
clean_sheets             int64
dtype: object
📊 season_name summary
• Season span: 00/01 → 99/00
• Number of unique seasons: 56
• Most frequent season: 24/25 (88,010 rows)

Counts per season:


season_name
00/01        8
01/02       18
02/03       37
03/04      109
04/05      204
05/06      409
06/07      859
07/08     1559
08/09     2557
09/10     3991
10/11     6118
11/12     8355
12/13    11261
13/14    14793
14/15    18329
15/16    23162
16/17    28226
17/18    33816
18/19    40279
19/20    44374
20/21    48680
2000         1
2001         3
2002         4
2003        17
2004        36
2005        52
2006       120
2007       215
2008       345
2009       616
2010       933
2011      1280
2012      2058
2013      2977
2014      3622
2015      4954
2016      5764
2017      7145
2018      8163
2019     10515
2020     12309
2021     13200
2022     17563
2023     20653
2024     22367
2025     21175
2026         4
21/22    61555
22/23    69497
23/24    77782
24/25    88010
25/26    20043
97/98        1
98/99        1
99/00        1
Name: count, dtype: int64

#### Summary

- **760,125 rows × 20 columns**  
- **Unit of observation:** one line per **player × season × competition × team**, aggregating match-level events for that slice (e.g., totals of appearances, goals, cards).  
- **Season coverage (`season_name`):** **1997/98 → 2025/26** (also includes some single-year labels `2000–2026`)  
- **Distinct seasons:** **56**  
- **Most frequent season:** **24/25** (88,010 rows)

**Columns**

- `player_id` — Player identifier.  
- `season_name` — Season label (e.g., `"2003/2004"`, `"24/25"`, sometimes single years like `"2025"`).  
- `competition_id` — Competition code.  
- `competition_name` — Competition name.  
- `team_id` — Team/club identifier for that season/competition.  
- `team_name` — Team/club name.  
- `nb_in_group` — Times named in the matchday squad.  
- `nb_on_pitch` — **Appearances** (played minutes > 0).  
- `goals` — Goals scored.  
- `assists` — Assists.  
- `own_goals` — Own goals.  
- `subed_in` — Times **substituted on**.  
- `subed_out` — Times **substituted off**.  
- `yellow_cards` — Yellow cards.  
- `second_yellow_cards` — Second yellows (leading to red).  
- `direct_red_cards` — Direct red cards.  
- `penalty_goals` — Goals scored from penalties.  
- `minutes_played` — Minutes played (may be `NaN` when unavailable).  
- `goals_conceded` — Goals conceded (primarily relevant for GKs/defensive lines).  
- `clean_sheets` — Clean sheets recorded.  

In [12]:
display(df.head(10))   

player_id season_name competition_id        competition_name  team_id  \
0    1058343       24/25           SCJ2  Supercoppa Primavera 2    17664   
1    1058343       24/25           IJ2B         Primavera 2 - B    17664   
2    1058343       24/25           ITJ7                Under 18    75868   
3    1058343       23/24           ITJF       Under 17 - Finals    48967   
4    1058343       23/24           ITJ5            Under 17 - B    48967   
5    1058343       23/24           ITJ7                Under 18    75868   
6     410547        2025            NO1             Eliteserien    23184   
7     410547        2025           NOPO                  NM-Cup    23184   
8     410547        2024            NO1             Eliteserien    23184   
9     410547        2024           NOPO                  NM-Cup    23184   

              team_name  nb_in_group  nb_on_pitch  goals  assists  own_goals  \
0   Frosinone Primavera            1            1    0.0        0          0   
1   Frosinone Primavera           15           12    0.0        0          0   
2        Inter Under 18           15           10    0.0        0          0   
3        Inter Under 17            1            0    0.0        0          0   
4        Inter Under 17           24           18    0.0        0          0   
5        Inter Under 18            7            2    0.0        0          0   
6  KFUM-Kameratene Oslo           15           15    3.0        1          0   
7  KFUM-Kameratene Oslo            4            4    2.0        1          0   
8  KFUM-Kameratene Oslo           30           28    0.0        1          0   
9  KFUM-Kameratene Oslo            6            6    2.0        0          0   

   subed_in  subed_out  yellow_cards  second_yellow_cards  direct_red_cards  \
0         0          1             0                    0                 0   
1         2          2             1                    0                 0   
2         6          1             1                    0                 0   
3         0          0             0                    0                 0   
4         2          6             1                    0                 0   
5         2          0             0                    0                 0   
6         1          9             2                    0                 0   
7         2          1             0                    0                 0   
8         9         10             1                    0                 0   
9         1          0             0                    0                 0   

   penalty_goals  minutes_played  goals_conceded  clean_sheets  
0              0             NaN               0             0  
1              0             NaN               0             0  
2              0             NaN               0             0  
3              0             NaN               0             0  
4              0             NaN               0             0  
5              0             NaN               0             0  
6              0           373.0               0             0  
7              0            95.0               0             0  
8              0             NaN               0             0  
9              0           281.0               0             0

#### To further examine how much of Messi carreer is considered in the `df`, we filter `player_id`

**IMPORTANT**\
The link between `player_id` and the name of the football player is in `player_profiles.csv`

##### **Lionel Messi**
- player_id = 28003
- player_name = “Lionel Messi”
- player_slug = “lionel-messi”
- Date of birth = 1987-06-24
- Current club = Inter Miami CF

In [9]:
# Apply filter using Messi's player_id
messi_id = 28003
messi_df = df[df["player_id"] == messi_id]

# Basic structure
n_rows, n_cols = messi_df.shape
print(f"Rows (records): {n_rows}")
print(f"Columns: {n_cols}")

# Season range instead of date range
season_min = messi_df["season_name"].min()
season_max = messi_df["season_name"].max()
print(f"Season span: {season_min} → {season_max}")

# Show first 10 rows
messi_df.head(89)

Rows (records): 89
Columns: 20
Season span: 03/04 → 24/25


player_id season_name competition_id       competition_name  team_id  \
384491      28003        2025           MLS1                    MLS    69261   
384492      28003        2025           USMX            Leagues Cup    69261   
384493      28003        2025           KLUB         Club World Cup    69261   
384494      28003       24/25            CCL            CONCACAF CC    69261   
384495      28003        2024           POUS       MLS Cup Playoffs    69261   
...           ...         ...            ...                    ...      ...   
384575      28003       04/05           ES3C       2ª B - Grupo III     2464   
384576      28003       04/05            ES1                 LaLiga      131   
384577      28003       04/05             CL  UEFA Champions League      131   
384578      28003       04/05            CDR           Copa del Rey      131   
384579      28003       03/04           ES3C       2ª B - Grupo III     2464   

             team_name  nb_in_group  nb_on_pitch  goals  assists  own_goals  \
384491  Inter Miami CF           20           19   19.0        8          0   
384492  Inter Miami CF            2            2    0.0        2          0   
384493  Inter Miami CF            4            4    1.0        0          0   
384494  Inter Miami CF            7            7    5.0        1          0   
384495  Inter Miami CF            3            3    1.0        1          0   
...                ...          ...          ...    ...      ...        ...   
384575  FC Barcelona B           17           17    6.0        1          0   
384576    FC Barcelona           16            7    1.0        0          0   
384577    FC Barcelona            4            1    0.0        0          0   
384578    FC Barcelona            1            1    0.0        0          0   
384579  FC Barcelona B            5            5    0.0        0          0   

        subed_in  subed_out  yellow_cards  second_yellow_cards  \
384491         2          0             2                    0   
384492         0          1             0                    0   
384493         0          0             1                    0   
384494         1          1             0                    0   
384495         0          0             0                    0   
...          ...        ...           ...                  ...   
384575         0          2             3                    0   
384576         7          0             0                    0   
384577         0          0             0                    0   
384578         0          1             0                    0   
384579         0          2             1                    0   

        direct_red_cards  penalty_goals  minutes_played  goals_conceded  \
384491                 0              0            85.0               0   
384492                 0              0             NaN               0   
384493                 0              0           360.0               0   
384494                 0              1           111.0               0   
384495                 0              0           270.0               0   
...                  ...            ...             ...             ...   
384575                 1              0           253.0               0   
384576                 0              0            76.0               0   
384577                 0              0             NaN               0   
384578                 0              0             NaN               0   
384579                 0              0             NaN               0   

        clean_sheets  
384491             0  
384492             0  
384493             0  
384494             0  
384495             0  
...              ...  
384575             0  
384576             0  
384577             0  
384578             0  
384579             0  

[89 rows x 20 columns]

### ⚽ **Note**
`season_name` coverage spans Messi’s entire career. Therefore, dataset 2 is prefect for comparing Messi’s performance with other players.  

#### We continue describing other variables which we consider KEY as possible filters:

### Competition
There are plenty of soccer players in the dataset, and we just want those players that represent a fair comparisson against Messi. We could narrow the soccer player list by filtering competition. We describe: `competition_id` & `competition_name`

In [7]:
# Competition coverage analysis
if "competition_id" not in df.columns or "competition_name" not in df.columns:
    raise KeyError("competition_id / competition_name columns not found in dataset")

# Number of unique competitions
n_competitions = df["competition_id"].nunique()

# Most frequent competition
most_freq_comp = df["competition_name"].value_counts().idxmax()
most_freq_comp_count = df["competition_name"].value_counts().max()

print("\n📊 competition summary")
print(f"• Number of unique competitions: {n_competitions}")
print(f"• Most frequent competition: {most_freq_comp} ({most_freq_comp_count:,} rows)")

unique_competitions = (
    df[["competition_id", "competition_name"]]
    .drop_duplicates()
    .sort_values("competition_name")
)
print("\nUnique competitions (sample):")
display(unique_competitions.head(20))


📊 competition summary
• Number of unique competitions: 1422
• Most frequent competition: FA Cup (13,098 rows)

Unique competitions (sample):


competition_id         competition_name
1695              PL2                   1 Liga
373               KR1                   1. HNL
753976            A4S               1. LL Sbg.
121              CHC1            1. Liga Gr. 1
26121             CR1  1. Liga Gr. 1 (- 11/12)
15894            CHC2            1. Liga Gr. 2
13884             CR2  1. Liga Gr. 2 (- 11/12)
514              CHC3            1. Liga Gr. 3
24191             CR3  1. Liga Gr. 3 (- 11/12)
7669             SUIP          1. Liga Playoff
299              CHC1            1. Liga gr. 1
7576             CHC2            1. Liga gr. 2
512              CHC3            1. Liga gr. 3
113420           FAR2                 1. deild
34693             NO2              1. divisjon
9439             CZ19              1. liga U19
86033             IS2                  1.Deild
918               RU2               1.Division
4102              DK2               1.Division
4104             DK2P     1.Division Promotion

### One good aproach is to filter by the Big Five divisions

🏆 **Big Five — Top Divisions**

- **Spain** → `ES1` → *LaLiga*  
- **Italy** → `IT1` → *Serie A*  
- **England** → `GB1` → *Premier League*  
- **France** → `FR1` → *Ligue 1*  
- **Germany** → `L1` → *Bundesliga*

The previous codes can be found in `team_details.csv`\
Dicision codes will be important for data curation in a later notebook.

###  Teams

We could narrow the soccer player list by filtering selected teams. We describe the variable `team_name`

In [8]:
if "team_name" not in df.columns:
    raise KeyError("team_name column not found in dataset")

# Number of unique teams
n_teams = df["team_name"].nunique()

# Most frequent team
most_freq_team = df["team_name"].value_counts().idxmax()
most_freq_team_count = df["team_name"].value_counts().max()

print("\n📊 team_name summary")
print(f"• Number of unique teams: {n_teams}")
print(f"• Most frequent team: {most_freq_team} ({most_freq_team_count:,} rows)")

team_counts = df["team_name"].value_counts()
print("\nCounts per team:")
display(team_counts.head(20))


📊 team_name summary
• Number of unique teams: 9404
• Most frequent team: Arsenal FC (1,816 rows)

Counts per team:


team_name
Arsenal FC                         1816
CR Flamengo                        1613
Ajax Amsterdam                     1541
CA River Plate                     1479
Sociedade Esportiva Palmeiras      1474
FC Porto                           1432
Manchester United                  1432
SL Benfica                         1426
CA Boca Juniors                    1419
Sport Club Corinthians Paulista    1415
São Paulo Futebol Clube            1413
PSV Eindhoven                      1407
Celtic FC                          1361
Sport Club Internacional           1338
Liverpool FC                       1321
Fluminense Football Club           1321
FC Barcelona                       1313
GNK Dinamo Zagreb                  1289
Club Brugge KV                     1284
Tottenham Hotspur                  1272
Name: count, dtype: int64

## One possibility is to select players coming from teams inside the big five during Messi activity window

The following info was provided by ChatGPT:
### Teams from the Big Five during Messi’s activity window (2003–2025)

### España — LaLiga (20)

Real Madrid · Barcelona · Atlético de Madrid · Sevilla · Valencia · Villarreal · Athletic Club · Real Sociedad · Real Betis · Celta de Vigo · Getafe · Osasuna · Espanyol · Mallorca · Levante · Rayo Vallecano · Málaga · Deportivo La Coruña · Real Valladolid · Alavés  

### Italia — Serie A (20)

Juventus · Inter · Milan · Roma · Lazio · Napoli · Fiorentina · Atalanta · Udinese · Torino · Sampdoria · Bologna · Genoa · Cagliari · Parma · Chievo Verona · Sassuolo · Hellas Verona · Palermo · Empoli  

### Inglaterra — Premier League (20)

Arsenal · Manchester United · Chelsea · Liverpool · Tottenham Hotspur · Manchester City · Everton · Aston Villa · Newcastle United · West Ham United · Southampton · Leicester City · Fulham · Crystal Palace · West Bromwich Albion · Stoke City · Sunderland · Brighton & Hove Albion · Wolverhampton Wanderers · AFC Bournemouth  

### Francia — Ligue 1 (18)

Paris Saint-Germain · Lyon · Marseille · Lille · Monaco · Rennes · Nice · Bordeaux · Saint-Étienne · Montpellier · Nantes · Toulouse · Lorient · Lens · Reims · Strasbourg · Metz · Caen  

### Alemania — Bundesliga (18)

Bayern München · Borussia Dortmund · Bayer Leverkusen · Schalke 04 · VfL Wolfsburg · VfB Stuttgart · Werder Bremen · Borussia Mönchengladbach · TSG Hoffenheim · 1. FSV Mainz 05 · SC Freiburg · Hertha BSC · FC Augsburg · Hannover 96 · 1. FC Köln · Hamburger SV · Eintracht Frankfurt · RB Leipzig  
